core/src/main/java/org/apache/calcite/rel/rel2sql/SqlImplementor.java

```diff
@@ -726,6 +726,31 @@ public SqlNode toSql(@Nullable RexProgram program, RexNode rex) {
           }
         }
         elseNode = caseNodeList.get(caseNodeList.size() - 1);
+
+        if (caseCall.getType().getSqlTypeName() == SqlTypeName.BOOLEAN
+            && !dialect.supportBooleanCaseWhen()) {
+          // Transformed when expressions of boolean type in SqlCase
+          // For example, given
+          //     CASE WHEN x > 1 THEN y > 1 ELSE y < 10 END
+          // Transformed:
+          //     CASE WHEN x > 1 THEN CASE WHEN y > 1 THEN 1 ELSE 0 END
+          // ELSE CASE WHEN y < 10 THEN 1 ELSE 0 END END = 1
```

Comment:
the comment looks wrong, is the last comparison supposed to be there?

Reply:
What's wrong with this, This guarantees the semantics of the bool expression
Regarding the tests expected in the comments below, Perhaps query1 in the test fulfills these, and I used true literal in the else expression

Reply:
Can you use some parentheses to make the order of evaluation clearer?
It's not obvious how the expression is parsed. 
I guess it's parsed as (CASE ... END) = 1
Now I am worried about a CASE expression that mixes boolean and non-boolean values.
How about something like `(CASE WHEN ... x > 1 ELSE 2 END)`?
Is the translation of this also preserving its semantics?

Reply:
good idea, i have added this part about parentheses.
In Oracle, would there be a difference between then and else expression statements, also in the default implementation of calcite this would throw the error "Illegal mixing of types in CASE or COALESCE statement"